File: src/leetcode/2025/12/Solution290.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <sstream>
using namespace std;

class Solution290
{
public:
    bool wordPattern(string pattern, string s)
    {
        unordered_map<char, string> charToWord;
        unordered_map<string, char> wordToChar;
        istringstream iss(s);
        string word;
        int index = 0;

        while (iss >> word)
        {
            if (index >= pattern.size())
            {
                return false;
            }

            char c = pattern[index];

            if (charToWord.count(c) && charToWord[c] != word)
            {
                return false;
            }
            if (wordToChar.count(word) && wordToChar[word] != c)
            {
                return false;
            }

            charToWord[c] = word;
            wordToChar[word] = c;
            index++;
        }

        return index == pattern.size();
    }
};

int main()
{
    Solution290 solution;

    string pattern1 = "abba";
    string s1 = "dog cat cat dog";
    cout << (solution.wordPattern(pattern1, s1) ? "true" : "false") << endl;
    // Expected output: true

    string pattern2 = "abba";
    string s2 = "dog cat cat fish";
    cout << (solution.wordPattern(pattern2, s2) ? "true" : "false") << endl;
    // Expected output: false

    string pattern3 = "aaaa";
    string s3 = "dog cat cat dog";
    cout << (solution.wordPattern(pattern3, s3) ? "true" : "false") << endl;
    // Expected output: false

    string pattern4 = "abba";
    string s4 = "dog dog dog dog";
    cout << (solution.wordPattern(pattern4, s4) ? "true" : "false") << endl;
    // Expected output: false

    return 0;
}

File: src/leetcode/2025/12/Solution290.java
import java.util.HashMap;
import java.util.Map;

public class Solution290 {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) {
            return false;
        }

        int[] charToIndex = new int[26];
        Map<String, Integer> wordToIndex = new HashMap<>();

        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String word = words[i];

            int patternIndex = charToIndex[c - 'a'];
            int wordIndex = wordToIndex.getOrDefault(word, 0);

            if (patternIndex != wordIndex) {
                return false;
            }

            charToIndex[c - 'a'] = i + 1;
            wordToIndex.put(word, i + 1);
        }

        return true;
    }

    public static void main(String[] args) {
        Solution290 solution = new Solution290();

        String pattern1 = "abba";
        String s1 = "dog cat cat dog";
        System.out.println(solution.wordPattern(pattern1, s1)); // true

        String pattern2 = "abba";
        String s2 = "dog cat cat fish";
        System.out.println(solution.wordPattern(pattern2, s2)); // false

        String pattern3 = "aaaa";
        String s3 = "dog cat cat dog";
        System.out.println(solution.wordPattern(pattern3, s3)); // false

        String pattern4 = "abba";
        String s4 = "dog dog dog dog";
        System.out.println(solution.wordPattern(pattern4, s4)); // false
    }
}
